crates/swc_ecma_lints/src/rules/prefer_const.rs

```diff
@@ -0,0 +1,257 @@
+                postinitialized: false,
+            },
+        );
+    }
+
+    fn collect_decl_pat(&mut self, initialized: bool, pat: &Pat) {
+        match pat {
+            Pat::Ident(BindingIdent { id, .. }) => {
+                self.add_var_meta(id, initialized);
+            }
+            Pat::Assign(AssignPat { left, .. }) => {
+                self.collect_decl_pat(initialized, left.as_ref());
+            }
+            Pat::Array(ArrayPat { elems, .. }) => {
+                elems.iter().for_each(|elem| {
+                    if let Some(elem) = elem {
+                        self.collect_decl_pat(initialized, elem);
+                    }
+                });
+            }
+            Pat::Object(ObjectPat { props, .. }) => {
+                props.iter().for_each(|prop| {
+                    match prop {
+                        ObjectPatProp::KeyValue(KeyValuePatProp { value, .. }) => {
+                            self.collect_decl_pat(initialized, value.as_ref());
+                        }
+                        ObjectPatProp::Assign(AssignPatProp { key, .. }) => {
+                            self.add_var_meta(key, initialized);
+                        }
+                        _ => {}
+                    };
+                });
+            }
+            _ => {}
+        }
+    }
+
+    fn consider_mutation_for_ident(&mut self, ident: &Ident, destructuring_assign: bool) {
+        let id = ident.to_id();
+
+        if let Some(var_meta) = self.vars_meta.get_mut(&id) {
+            if destructuring_assign && !var_meta.initialized {
+                var_meta.destructuring_assign = true;
+                var_meta.span = ident.span;
+
+                return;
+            }
+
+            if var_meta.initialized || self.block_depth > var_meta.block_depth {
+                self.vars_meta.remove(&id);
+
+                return;
+            }
+
+            var_meta.postinitialized = true;
+        }
+    }
+
+    fn consider_mutation(&mut self, pat: &Pat, destructuring_assign: bool) {
+        match pat {
+            Pat::Ident(BindingIdent { id, .. }) => {
+                self.consider_mutation_for_ident(id, destructuring_assign);
+            }
+            Pat::Array(ArrayPat { elems, .. }) => elems.iter().for_each(|elem| {
+                if let Some(pat) = elem {
+                    self.consider_mutation(pat, destructuring_assign);
+                }
+            }),
+            Pat::Object(ObjectPat { props, .. }) => {
+                props.iter().for_each(|prop| match prop {
+                    ObjectPatProp::KeyValue(KeyValuePatProp { value, .. }) => {
+                        self.consider_mutation(value.as_ref(), true);
+                    }
+                    ObjectPatProp::Assign(AssignPatProp { key, .. }) => {
+                        self.consider_mutation_for_ident(key, true);
+                    }
+                    _ => {}
```

Comment:
I'm not sure, but we may need to consider rest patterns

Reply:
Yes, you're right. eslint rule considers rest declaration